src/Microsoft.AspNetCore.Blazor/RenderTree/RenderTreeDiffBuilder.cs

```diff
@@ -219,7 +271,7 @@ private static int NextSiblingIndex(RenderTreeFrame frame, int frameIndex)
                             var newFrameAttributesEndIndexExcl = GetAttributesEndIndexExclusive(newTree, newFrameIndex);
 
                             // Diff the attributes
-                            AppendDiffEntriesForRange(
+                            AppendAttributeDiffEntriesForRange(
```

Comment:
Regarding perf, I haven't measured it, but do you have a sense of how the old and new mechanisms would compare in the case where the attributes are unchanged? I'd think that 95%+ of the time, the attribute names are at least unchanged (same set in the same order). Conditional or procedurally-generated attribute names are relatively rare.

For the old method, I expected (but didn't verify) that in virtually all cases the attribute name comparisons would reduce down to reference equality checks without needing to iterate through the characters of the strings, since they'd come from the same compile-time constants. Whereas for the new method it will need to compute a hash code for each attribute name, which at least involves iterating through all the characters.

I know there's a risk of over-thinking the optimisation here, but given that in the vast majority of cases the set of attribute names will be unchanged, do you think it might be worth doing a first-pass check here where we establish whether the set of attribute names are unchanged (by looping through them, expecting the string comparisons to be reference-equality checks in most cases), and if they are, falling back on the earlier `AppendDiffEntriesForRange` logic to avoid the hashcode calculations?

Reply:
Or a hybrid like where we start with the old approach, but switch to the new one when we find the first changed attribute name:

    for(var index = start; index < endExcl; index++)
    {
        if (oldAttributeName[index] == newAttributeName[index])
        {
            AppendDiffEntriesForFramesWithSameSequence(... index ...);
        }
        else
        {
            // Something's changed. Process the remaining attributes using the hash lookups
            AppendAttributeDiffEntriesForRange(index, endExcl, ...);
            break;
        }    
    }

Reply:
Actually that approach might get more messy when the two sets have different numbers of entries. I'll stop speculating now as I'm sure you will have a view on how you'd prefer to handle this, if at all.

Reply:
I'll take a look at this and see if it makes a significant difference. 

One idea that I had is to do a fast path for single attributes since a single entry can't ever be in a different order 👍 

If your main concern is the cost of hashing, we could build a really simple equality comparer that just uses the first character or length. I'm interested in trying this out because I want to see if this is useful for routing 😆

Note, we still need to diff the attribute values in all cases.